src/network/src/esp/src/TCPServer.cpp

```diff
@@ -0,0 +1,109 @@
+#include "TCPServer.h"
+#include "LockGuard.h"
+#include "NetworkConfig.h"
+#include "SocketFactory.h"
+#include "Task.h"
+#include "lwip/sockets.h"
+
+constexpr uint8_t gs_LOOP_RATE = 10;
+
+static void serverReceiveTask(void* context) {
+
+    if (context != nullptr) {
+        while (true) {
+            static_cast<TCPServer*>(context)->receiveTask();
+            Task::delay(gs_LOOP_RATE);
+        }
+    }
+}
+
+TCPServer::TCPServer(ILogger& logger) :
+    m_serverTask("server_socket", tskIDLE_PRIORITY + 1, serverReceiveTask, this),
+    m_socket(NO_SOCKET),
+    m_logger(logger),
+    m_serverMutex(10) {
+    m_isBusy = false;
+
+    CircularBuff_init(&m_circularBuffer, m_buffer.data(), m_buffer.size());
+
+    m_serverTask.start();
+}
+
+bool TCPServer::send(const uint8_t* data, uint16_t length) {
+    (void)data;
+    (void)length;
+    m_logger.log(LogLevel::Error, "Sending not supported on TCP server");
+    return false;
+}
+
+bool TCPServer::receive(uint8_t* data, uint16_t length) {
+    if (data == nullptr || length > m_buffer.size()) {
+        m_logger.log(LogLevel::Warn, "Invalid parameters for Server receive");
+        return false;
+    }
+
+    LockGuard lock = LockGuard(m_serverMutex);
+    CircularBuff_get(&m_circularBuffer, data, length);
```

Comment:
What happens if we try to get more than what is currently in the buffer? You could check what is in the HM UART driver where this call is blocked on a signal and unblocked whenever data is added to the buffer

Reply:
True, forgot to add this check